### PR TITLE
AWS Bedrock Integration

### PR DESCRIPTION
**Pull request by daftops**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: not applicable

## Description
Send question to AWS LLM by using AWS API:
Reference:
- https://docs.aws.amazon.com/bedrock/latest/APIReference/API_runtime_InvokeModel.html
- https://docs.aws.amazon.com/bedrock/latest/userguide/api-methods-run-inference.html

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by daftops** [comment]
Please note minimal Boto3 version must be [1.28.57](https://github.com/boto/boto3/blob/master/.changes/1.28.57.json)

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    contentType = 'application/json'
+
+    response = bedrock.invoke_model(body=body, modelId=modelId, accept=accept, contentType=contentType)
+
+    response_body = json.loads(response.get('body').read())
+
+    return CommandResults(
+        outputs_prefix='AWS.Bedrock',
+        readable_output=f'AWS Bedrock API returns with HTTP Code '
+                        f'{response.get("ResponseMetadata").get("HTTPStatusCode")}\n{response_body.get("completion")}',
```
Comment (sapirshuker) [inline Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py:25]:
```suggestion
                        f'{response.get("ResponseMetadata",{}).get("HTTPStatusCode",{})}\n{response_body.get("completion")}',
```

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    params = demisto.params()
+    command = demisto.command()
+    args = demisto.args()
+
+    # AWS Config
+    aws_default_region = params.get('defaultRegion')
+    aws_role_arn = params.get('roleArn')
+    aws_role_session_name = params.get('roleSessionName')
+    aws_role_session_duration = params.get('sessionDuration')
+    aws_role_policy = None
```
Comment (sapirshuker) [inline Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py:50]: Why the aws_role_policy is None?

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+This README contains the full documentation for your integration.
+
+You auto-generate this README file from your integration YML file using the `demisto-sdk generate-docs` command.
```
Comment (sapirshuker) [inline Packs/AWS_Bedrock/Integrations/AwsBedrock/README.md:3]: Please generate a README.md file for your integration/script.
To do that, please run the `demisto-sdk generate-docs` command (use the -e flag for using a commad_examples file) - for more information please see: https://xsoar.pan.dev/docs/concepts/demisto-sdk#generate-docs.

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock_test.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    Given:
+    - A basic question.
+    When:
+    - Calling aws-bedrock-ask command.
+    Then:
+    - Ensure that the result is correct and JSON.
+    """
+    from AwsBedrock import ask_command
+
+    bedrock = botocore.session.get_session().create_client('bedrock-runtime', region_name='us-east-1')
```
Comment (sapirshuker) [inline Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock_test.py:24]: We don't want to use in any external functions in the unit test, instead we are using mock please see [here](https://xsoar.pan.dev/docs/integrations/unit-testing#mocking) 
[You can see exmples in our repo to mocking unit test 
Packs/AWS-Route53/Integrations/AWSRoute53/AWSRoute53_test.py](https://github.com/demisto/content/blob/2fab65a0ecc88352fa7e3585e5a65554215401db/Packs/AWS-Route53/Integrations/AWSRoute53/AWSRoute53_test.py#L4)

**Review by sapirshuker — CHANGES_REQUESTED** [review]
Hey, @daftops
Thank you for your contribution!
Good work :)
Please see my comments
Feel free to reach out to me with any questions - I'm available here or on slack :)
Thanks again
In addition, please note that logo should match the requirements (120x50 px PNG, transparent background, <10kB in Size)

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  - name: aws-bedrock-ask
+    arguments:
+    - name: question
+      description: The question to be sent to AWS Bedrock.
+      defaultValue: "65"
+      type: textArea
+    - name: model
+      description: Model to use in AWS Bedrock.
+      defaultValue: anthropic.claude-v2
+      type: textArea
```
Comment (sapirshuker) [inline Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml:86]:
```suggestion
```

**Comment by sapirshuker** [comment]
Hey @daftops thank you again for the great job you did on this contribution!
After discussing this contribution with our product team, Most of our AWS packs so far are xsoar support level and we wise to keep it that way.
We will happily accept your contribution and continue with the contribution processes, but it first needs to be converted to xsoar support level.
Xsoar support level requires higher code standards, unit tests, and a test playbook.
We can assist you in bringing the contribution to xsoar support level if you wish to continue with it
In case you decide you don’t want to make the extra effort we will develop this as an internal pack in the future.
Please let me know what is your decision, Thanks a lot!

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py (hunk 2 above)
Comment (upstart-swiss) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py:50]: This is declared in the AWSClient as a required argument (palo developed). Other areas such as lambda where Philippe linked to declare it as None.

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+script:
+  script: ''
+  type: python
+  commands:
+  - name: aws-bedrock-ask
+    arguments:
+    - name: question
+      description: The question to be sent to AWS Bedrock.
+      defaultValue: "65"
+      type: textArea
```
[the remark this answers is not in the document]
Comment (upstart-swiss) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml:79]: I think questions can be multiline, so a textArea likely makes more sense. Is there a reason you're opposed to one?

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/test_data/aws_bedrock_response.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+        "ResponseMetadata": {
+            "RequestId": "3fd51457-f34a-4d9c-8ef8-a207a3acdac3",
```
[the remark this answers is not in the document]
Comment (upstart-swiss) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/test_data/aws_bedrock_response.json:3]: This isn't a secret, it's an example GUID that aws would provide in its response.

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    - name: model
+      description: Model to use in AWS Bedrock.
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml:81]: Indeed doing this modification

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  required: false
+- display: Use system proxy settings
+  name: proxy
+  type: 8
+  required: false
+script:
+  script: ''
+  type: python
+  commands:
+  - name: aws-bedrock-ask
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml]: I modified it accordingly to your comment

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    arguments:
+    - name: question
+      description: The question to be sent to AWS Bedrock.
+      defaultValue: "65"
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml]: Legacy code, it has been removed

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
+from CommonServerPython import *  # noqa # pylint: disable=unused-wildcard-import
+
+
+def ask_command(bedrock, args):
+
+    question = args.get('question')
+    body = json.dumps({
+        "prompt": f"\n\nHuman:{question}\n\nAssistant:",
+        "max_tokens_to_sample": 30,
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.py]: I modified to be an argument please note the following

> Maximum length (max_tokens_to_sample) – Specify the maximum number of tokens to use in the generated response. We recommend a limit of 4,000 tokens for optimal performance.

According to https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-claude.html#model-parameters-claude-length

So response could not be larger than 30 based on the documentation.

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (sapirshuker) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml:79]: My mistake, please ignore from this comment

**Comment by MLainer1** [comment]
Hi @daftops, please remember to resolve the comments that you already fixed so we will see when you are finished and waiting for a review, thanks!

File: Packs/AWS_Bedrock/pack_metadata.json [hunk 12]
```diff
@@ -0,0 +1,23 @@
+{
+    "name": "AWS Bedrock",
+    "description": "Send command to AWS Bedrock",
+    "support": "community",
+    "currentVersion": "1.0.0",
+    "author": "Philippe Laporte",
+    "url": "https://live.paloaltonetworks.com/t5/cortex-xsoar-discussions/bd-p/Cortex_XSOAR_Discussions",
+    "email": "",
+    "categories": [
+        "Analytics & SIEM"
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/pack_metadata.json:10]: no it is not, I will modify accordingly

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml [hunk 13]
```diff
@@ -0,0 +1,97 @@
+commonfields:
+  id: AWS Bedrock
+  version: -1
+name: AWS Bedrock
+display: AWS Bedrock
+category: Utilities
+description: This is Amazon Bedrock integration.
```
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock.yml]: I updated my description by using the following AWS integrations description supported by PAN:
- https://github.com/demisto/content/blob/master/Packs/AWS-Route53/Integrations/AWSRoute53/AWSRoute53.yml#L68-L69
- https://github.com/demisto/content/blob/master/Packs/AWS-Lambda/Integrations/AWS_Lambda/AWS_Lambda.yml#L88

File: Packs/AWS_Bedrock/pack_metadata.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (daftops) [reply Packs/AWS_Bedrock/pack_metadata.json]: I updated according to category used by other AWS pack/integration such as:

AWS Route 53:
- https://github.com/demisto/content/blob/master/Packs/AWS-Route53/Integrations/AWSRoute53/AWSRoute53.yml#L1
- https://github.com/demisto/content/blob/master/Packs/AWS-Route53/pack_metadata.json#L11

AWS Lambda:
- https://github.com/demisto/content/blob/master/Packs/AWS-Lambda/pack_metadata.json#L11
https://github.com/demisto/content/blob/master/Packs/AWS-Lambda/Integrations/AWS_Lambda/AWS_Lambda.yml#L1

File: Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock_test.py (hunk 4 above)
Comment (daftops) [reply Packs/AWS_Bedrock/Integrations/AwsBedrock/AwsBedrock_test.py]: I just modified, let me know if it matches what you expect

**Review by MLainer1 — CHANGES_REQUESTED** [review]
Nice work! Let's schedule a demo to see the new pack in action. I'll reach over slack to schedule it.

File: Packs/AWS_Bedrock/pack_metadata.json [hunk 14]
```diff
@@ -0,0 +1,23 @@
+{
+    "name": "AWS Bedrock",
+    "description": "Send command to AWS Bedrock",
+    "support": "community",
```
Comment (MLainer1) [inline Packs/AWS_Bedrock/pack_metadata.json:4]: Please change the support level to xsoar

**Comment by MLainer1** [comment]
Hi @daftops, thank you for your time in the demo. Those are the points that were raised in the demo:
- Unify the credentials to one object.
- Create a more informative error when the credentials isn't valid.
- Let's check if the connection should fail when changing the Role session Name.
- Change the second argument's description.
- Make the question argument to be mandatory.
- Change the outputs prefix to AWS.Bedrock.Response .
- Change the human readable.
- Change the name of the command to aws_bedrock_ask_question.
- Add a test playbook to test the integration.
- Change the support level of the integration to xsoar support.
After you make the requested changes, please record a video of running the command in the playground, especially with other models, just to check the integrity of the command, and the run of the test playbook.

**Comment by daftops** [comment]
> Hi @daftops, thank you for your time in the demo. Those are the points that were raised in the demo:
> 
> * Unify the credentials to one object.
> * Create a more informative error when the credentials isn't valid.
As discussed during the meeting can you point me AWS examples where there's informative error.

> * Let's check if the connection should fail when changing the Role session Name.
As explained, I'm not AWS expert so I don't know where can I find this info but I strongly believe an Role Session name is informative

I'll work next week on the point below:
> * Change the second argument's description.
> * Make the question argument to be mandatory.
> * Change the outputs prefix to AWS.Bedrock.Response .
> * Change the human readable.
> * Change the name of the command to aws_bedrock_ask_question.
> * Add a test playbook to test the integration.
> * Change the support level of the integration to xsoar support.

> After you'll make the requested changes, please record a video of running the command in the playground, especially with other models, just to check the integrity of the command, and the run of the test playbook.

Thank you for your time @sapirshuker  @MLainer1

**Comment by MLainer1** [comment]
Hi @daftops, Your contribution was impacted by changes in the git history of the content repo, and you see many file changes that are not related to your PR. Our team is currently working on a solution, and we hope that it will be resolved soon.

**Comment by MLainer1** [comment]
Hello @daftops,

Last Thursday, December 20th, there was an update in the content repo. This is why your PR looks like there were a lot of commits and files changed.
To fix this issue please do the following:
1. Save all of your changes locally or in a new branch.
2. Enter your forked repository page in GitHub.
3. At the top of the page, you will probably see something like this, please ensure you are in the master branch.
image

5. Click on "Sync fork" button on the top right side of the page.
image

6. The following message will appear (This branch has conflicts that must be resolved).

7. Click on the button "Discard commits". Note that for this action to work, you should have permission to do a force push in the forked repo. If you don't have one, ask the forked repo Owner to do that.
8. Enter your branch and repeat the same steps (4-6).
9. Check your PR, and see if your changes have been saved. If yes - you finished the update process.
10. If your PR does not show your changes, open a new PR with the changes that you saved at the beginning of the process.
11. In case you opened a new PR please close the old one, and update your reviewer with the new PR link.
12. Please paste on the new open PR the URL to your previous PR so our team member can continue the review from the same place.

Thank you very much for your contribution and your understanding of the situation.
Have a nice holiday season!

**Comment by daftops** [comment]
> Last Thursday, December 20th, there was an update in the content repo. This is why your PR looks like there were a lot of commits and files changed. To fix this issue please do the following:
> 
> 1. Save all of your changes locally or in a new branch.
> 2. Enter your forked repository page in GitHub.
> 3. At the top of the page, you will probably see something like this, please ensure you are in the master branch.
> 
> image
> 5. Click on "Sync fork" button on the top right side of the page.
> 
> image
> 6. The following message will appear (This branch has conflicts that must be resolved).
> 7. Click on the button "Discard commits". Note that for this action to work, you should have permission to do a force push in the forked repo. If you don't have one, ask the forked repo Owner to do that.
> 8. Enter your branch and repeat the same steps (4-6).
> 9. Check your PR, and see if your changes have been saved. If yes - you finished the update process.
> 10. If your PR does not show your changes, open a new PR with the changes that you saved at the beginning of the process.
> 11. In case you opened a new PR please close the old one, and update your reviewer with the new PR link.
> 12. Please paste on the new open PR the URL to your previous PR so our team member can continue the review from the same place.
> 
> Thank you very much for your contribution and your understanding of the situation. Have a nice holiday season!

New PR https://github.com/demisto/content/pull/31852

**Pull request merged**